### ISSUE TITLE
First-class API to replace tokens while copying files

### ISSUE THREAD

**Issue by lptr** [gradle/gradle#16606]
Substituting values for tokens in certain resource templates is a very common use case. Currently the recommended way to do it is like so:

```groovy
processIntegrationTestResources {
    inputs.files(configurations.corePersistenceEntitiesJar)
            .withPropertyName('configurations.corePersistenceEntitiesJar')
            .withPathSensitivity(PathSensitivity.RELATIVE)
    inputs.files(configurations.xxxPersistenceEntitiesJar)
            .withPropertyName('configurations.xxxPersistenceEntitiesJar')
            .withPathSensitivity(PathSensitivity.RELATIVE)

    eachFile {
        if (it.sourcePath == 'META-INF/persistence.xml') {
            it.filter(ReplaceTokens, tokens: [
                    corePersistenceEntitiesJar: configurations.corePersistenceEntitiesJar.asPath,
                    xxxPersistenceEntitiesJar: configurations.xxxPersistenceEntitiesJar.asPath,
                    classesDir: sourceSets.main.java.outputDir.path
            ])
        }
    }
}
```

There are a number of problems with this:

- the code is very verbose and somewhat counter-intuitive (we don't want to process each file for example)
- it is easy to forget to manually register inputs, or to register them incorrectly (note that the names of the tokens are not captured here as inputs at all)
- you need to reference a class from Ant (`ReplaceTokens`)

Clearly, we can do better here. Something like this might be an option:

```groovy
processIntegrationTestResources {
    replaceTokens("META-INF/persistence.xml",
        corePersistenceEntitiesJar: configurations.corePersistenceEntitiesJar.map { it.asPath },
        xxxPersistenceEntitiesJar: configurations.xxxPersistenceEntitiesJar.map { it.asPath },
        classesDir: sourceSets.main.java.classesDirectory.map { it.path }
    )
}
```

This way we could properly capture all the inputs without the user having to specify anything, and by using mapped `Provider`s we'd keep track of dependencies on the inputs as well.

---
cc: @gradle/execution

**Comment by melix** [comment]
+1

would be great if we had some better API for _binary filtering_ too.

**Comment by runningcode** [comment]
Would it also be possible to specify the timestamp on the resulting file?

Binary filtering sounds great to solve a lot of caching issues.

**Comment by lptr** [comment]
Could you clarify what you mean by binary filtering? It might make sense to open a separate issue, too, if it's a sufficiently different problem. We can deliver them together even so.

**Comment by lptr** [comment]
I was just made aware of `AbstractCopyTask.expand()`, which has somewhat different characteristics:

- it involves executing Groovy code as part of the substitution, which might be problematic because of Groovy itself, or because it makes the replacements Turing-complete,
- it applies to all files within a `CopySpec` (can work around by declaring files to process as a different `CopySpec`, which might actually be good practice),
- it also does not track dependencies.

It would probably make sense to extend `expand()`'s logic to handle `Provider`s as inputs. I think scoping to a given set of files could actually work better via a separate `CopySpec`.

**Comment by melix** [comment]
There's an existing issue for binary filtering, which has been marked as solved: https://github.com/gradle/gradle/issues/4010

I have personally faced this problem several times, that the only available filtering is on text.

**Comment by lptr** [comment]
It seems to me that the best way to handle binary filters would be to reopen #4010.

**Comment by lptr** [comment]
One other major issue with `expand()` is that its configuration is currently not tracked as an input. :/

**Comment by melix** [comment]
unstale

**Comment by ysb33r** [comment]
> This way we could properly capture all the inputs without the user having to specify anything, and by using mapped Providers we'd keep track of dependencies on the inputs as well.

Definitely! A common problem is that the tokens cannot be lazy-evaluated, and the only current workaround is to use `afterEvaluate` to set them.

In your example you seem to have `replaceTokens(String path, Provider<String>... tokens)` as the signature. We need  the ability to specify `startTOken` and `endToken` as well.
